### PR TITLE
replica exec impl execute commands and instances

### PR DESCRIPTION
**Pull request by liubaohai**
# Description           

## Type of change       

- **New feature**       

# Checklist:

- [x] **Self-review**: I have performed a **self-review** of my own code
- [x] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [ ] **Doc**:         I have made corresponding changes to the **documentation**
- [ ] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes

### CODE REVIEW

File: components/epaxos/src/replica/exec.rs [hunk 1]
```diff
@@ -48,4 +56,84 @@ impl<Engine> Replica<Engine> {
...
+    fn execute_commands(&mut self, inst: &Instance) -> Result<Vec<ExecuteResult>, Error> {
+        let mut rst = Vec::new();
+        for cmd in inst.cmds.iter() {
+            if OpCode::NoOp as i32 == cmd.op {
+                rst.push(ExecuteResult::Success);
+            } else if OpCode::Set as i32 == cmd.op {
+                rst.push(ExecuteResult::Success);
+                self.snapshot.set_kv(&cmd.key, &cmd.value)?;
+            } else if OpCode::Get as i32 == cmd.op {
+                match self.snapshot.get_kv(&cmd.key) {
```
Comment (drmingdrmer) [inline components/epaxos/src/replica/exec.rs]: 这个snapshot改个名字吧. storage咋样?

Comment (liubaohai) [reply components/epaxos/src/replica/exec.rs]: 嗯，可以

File: components/epaxos/src/replica/exec.rs [hunk 2]
```diff
@@ -48,4 +56,84 @@ impl<Engine> Replica<Engine> {
...
+                    return Err(e);
+                }
+            };
+            rst.push(inst.instance_id.unwrap());
+        }
+
+        match self.snapshot.trans_commit() {
+            // TODO send replys to client
+            Ok(_) => {}
+            Err(_) => self.snapshot.trans_rollback()?,
```
Comment (pengsven) [inline components/epaxos/src/replica/exec.rs]:
```suggestion
            Err(e) => {
                self.snapshot.trans_rollback()?;
                return Err(e);
            },
```

File: components/epaxos/src/replica/exec.rs [hunk 3]
```diff
@@ -48,4 +56,84 @@ impl<Engine> Replica<Engine> {
...
+                        SnapError::NotFound => rst.push(ExecuteResult::NotFound),
+                        _ => return Err(Error::from(e)),
+                    },
+                }
+            } else {
+                return Err(Error::CmdNotSupport(format!("{:?}", cmd.op)));
+            }
+        }
+
+        let mut new_inst = inst.clone();
```
[the remark this answers is not in the document]
Comment (liubaohai) [reply components/epaxos/src/replica/exec.rs]: new_inst.executed 这个字段需要修改

**Pull request merged**